toxcore/group_moderation.c

```diff
@@ -0,0 +1,844 @@
+        return false;
+    }
+
+    if (memcmp(hash, creds->hash, MOD_SANCTION_HASH_SIZE) != 0) {
+        LOGGER_WARNING(moderation->log, "Invalid credentials hash");
+        return false;
+    }
+
+    if (creds->checksum != sanctions_creds_get_checksum(creds)) {
+        LOGGER_WARNING(moderation->log, "Invalid credentials checksum");
+        return false;
+    }
+
+    if (moderation->shared_state_version > 0) {
+        if ((creds->version < moderation->sanctions_creds.version)
+                && !(creds->version == 0 && moderation->sanctions_creds.version == UINT32_MAX)) {
+            LOGGER_WARNING(moderation->log, "Invalid version");
+            return false;
+        }
+    }
+
+    if (!crypto_signature_verify(creds->sig, hash, MOD_SANCTION_HASH_SIZE, creds->sig_pk)) {
+        LOGGER_WARNING(moderation->log, "Invalid signature");
+        return false;
+    }
+
+    return true;
+}
+
+bool sanctions_list_check_integrity(const Moderation *moderation, const Mod_Sanction_Creds *creds,
+                                    const Mod_Sanction *sanctions, uint16_t num_sanctions)
+{
+    for (uint16_t i = 0; i < num_sanctions; ++i) {
+        if (!sanctions_list_validate_entry(moderation, &sanctions[i])) {
+            LOGGER_WARNING(moderation->log, "Invalid entry");
+            return false;
+        }
+    }
+
+    return sanctions_creds_validate(moderation, sanctions, creds, num_sanctions);
+}
+
+/** @brief Removes index-th sanction list entry.
+ *
+ * New credentials will be validated if creds is non-null.
+ *
+ * Returns true on success.
+ */
+non_null(1) nullable(3)
+static bool sanctions_list_remove_index(Moderation *moderation, uint16_t index, const Mod_Sanction_Creds *creds)
+{
+    if (index >= moderation->num_sanctions) {
+        return false;
+    }
+
+    const uint16_t new_num = moderation->num_sanctions - 1;
+
+    if (new_num == 0) {
+        if (creds != nullptr) {
+            if (!sanctions_creds_validate(moderation, nullptr, creds, 0)) {
+                return false;
+            }
+
+            moderation->sanctions_creds = *creds;
+        }
+
+        sanctions_list_cleanup(moderation);
+
+        return true;
+    }
+
+    /** Operate on a copy of the list in case something goes wrong. */
+    const size_t old_count =  moderation->num_sanctions;
+    Mod_Sanction *sanctions_copy = (Mod_Sanction *)calloc(old_count, sizeof(Mod_Sanction));
+
+    if (sanctions_copy == nullptr) {
+        return false;
+    }
+
+    memcpy(sanctions_copy, moderation->sanctions, old_count * sizeof(Mod_Sanction));
+
+    if (index != new_num) {
+        sanctions_copy[index] = sanctions_copy[new_num];
+    }
+
+    Mod_Sanction *new_list = (Mod_Sanction *)realloc(sanctions_copy, new_num * sizeof(Mod_Sanction));
+
+    if (new_list == nullptr) {
+        free(sanctions_copy);
+        return false;
+    }
+
+    if (creds != nullptr) {
+        if (!sanctions_creds_validate(moderation, new_list, creds, new_num)) {
```

Comment:
It might be neater to pass this the old list + the new entry to validate the credentials against? That way you don't have to jump through all these hoops to resize a copy of the list so you can drop it if credential validation fails.

Reply:
`sanction_creds_validate()` is validating the sanctions credentials rather than a single entry. It needs to see the entire sanctions list as a whole to do validation, which includes validating a signature of a hash of every sanctions entry along with the current version.

Reply:
Wouldn't passing both the current list and the new entry be the full sanctions list?